src/admm/acopf_admm_gpu.jl

```diff
@@ -372,34 +381,40 @@ function admm_restart(env::AdmmEnv; iterlim=800, scale=1e-4)
 
     u_curr = zeros(mod.nvar)
     copyto!(u_curr, sol.u_curr)
-
-    rateA_nviols, rateA_maxviol, rateC_nviols, rateC_maxviol = check_linelimit_violation(data, u_curr)
-    @printf(" ** Line limit violations\n")
-    @printf("RateA number of violations = %d (%d)\n", rateA_nviols, mod.nline)
-    @printf("RateA maximum violation    = %.2f\n", rateA_maxviol)
-    @printf("RateC number of violations = %d (%d)\n", rateC_nviols, mod.nline)
-    @printf("RateC maximum violation    = %.2f\n", rateC_maxviol)
-
-    @printf(" ** Time\n")
-    @printf("Generator = %.2f\n", time_gen)
-    @printf("Branch    = %.2f\n", time_br)
-    @printf("Bus       = %.2f\n", time_bus)
-    @printf("Total     = %.2f\n", time_gen + time_br + time_bus)
-
     objval = sum(data.generators[g].coeff[data.generators[g].n-2]*(data.baseMVA*u_curr[mod.gen_start+2*(g-1)])^2 +
                  data.generators[g].coeff[data.generators[g].n-1]*(data.baseMVA*u_curr[mod.gen_start+2*(g-1)]) +
                  data.generators[g].coeff[data.generators[g].n]
-                 for g in 1:mod.ngen)
-    @printf("Objective value = %.6e\n", objval)
+                 for g in 1:mod.ngen)::Float64
+    sol.objval = objval
+
+
+    rateA_nviols, rateA_maxviol, rateC_nviols, rateC_maxviol = check_linelimit_violation(data, u_curr)
+    if par.verbose > 0
+        @printf(" ** Line limit violations\n")
+        @printf("RateA number of violations = %d (%d)\n", rateA_nviols, mod.nline)
+        @printf("RateA maximum violation    = %.2f\n", rateA_maxviol)
+        @printf("RateC number of violations = %d (%d)\n", rateC_nviols, mod.nline)
+        @printf("RateC maximum violation    = %.2f\n", rateC_maxviol)
+
+        @printf(" ** Time\n")
+        @printf("Generator = %.2f\n", time_gen)
+        @printf("Branch    = %.2f\n", time_br)
+        @printf("Bus       = %.2f\n", time_bus)
+        @printf("Total     = %.2f\n", time_gen + time_br + time_bus)
+
+        @printf("Objective value = %.6e\n", objval)
+    end
+    return
 end
 
-function admm_rect_gpu(case; iterlim=800, rho_pq=400.0, rho_va=40000.0, scale=1e-4,
-                       use_gpu=false, use_polar=true, gpu_no=1)
-    TArray = ifelse(use_gpu, CuArray, Array)
+function admm_rect_gpu(case, ::Type{VT}; iterlim=800, rho_pq=400.0, rho_va=40000.0, scale=1e-4,
+                       use_gpu=false, use_polar=true, gpu_no=1) where VT
```

Comment:
Could we take verbosity option so that we can control it without modifying code?